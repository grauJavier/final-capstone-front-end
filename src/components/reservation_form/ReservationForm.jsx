import { useEffect, useState } from 'react';
import { useForm } from 'react-hook-form';
import { useSelector, useDispatch } from 'react-redux';
import fetchCities from '../../redux/reservation_form/actions/fetchCities';
import fetchPlaces from '../../redux/reservation_form/actions/fetchPlacesByCityId.js';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import { selectUserId } from '../../redux/user/userSlice.js';
import createReservation from '../../redux/my_reservations/actions/createReservation.js';

const ReservationForm = () => {
  const {
    handleSubmit
  } = useForm();

  const cities = useSelector((state) => state.reservation.cities);
  const places = useSelector((state) => state.reservation.places);
  const status = useSelector((state) => state.reservation.status);
  const dispatch = useDispatch();

  const [selectedCity, setSelectedCity] = useState('');
  const [selectedDate, setSelectedDate] = useState(null);
  const [selectedPlace, setSelectedPlace] = useState('');

  const userId = useSelector(selectUserId);

  useEffect(() => {
    const searchParams = new URLSearchParams(location.search);
    const selectedCityFromQuery = searchParams.get('city_id');
    const selectedPlaceFromQuery = searchParams.get('place_id');

    if (selectedCityFromQuery) {
      setSelectedCity(selectedCityFromQuery);
    }
    if (selectedPlaceFromQuery) {
      setSelectedPlace(selectedPlaceFromQuery);
    }
  }, [location.search]);

  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchCities());
    }
  }, [status, dispatch]);

  useEffect(() => {
    if (selectedCity) {
      const selectedCityObj = cities.find((city) => city.id === parseInt(selectedCity, 10));
      if (selectedCityObj) {
        dispatch(fetchPlaces(selectedCityObj.id));
      }
    }
  }, [selectedCity, dispatch, cities]);

  const onSubmit = () => {
    // Send the data to the server
    const data = {
      user_id: userId,
      place_id: selectedPlace,
      schedule_date: selectedDate.toISOString().split('T')[0],
    };

    dispatch(createReservation(data));
  };

  const today = new Date();
  today.setHours(0, 0, 0, 0);

  return (
    <section className="lg:ml-[15%] bg-golden-img bg-cover min-h-screen flex flex-col items-center justify-center p-4">
      <div className="container flex flex-col items-center justify-center sm:p-24 sm:pb-40">
        <h2 className="secondary-font text-4xl font-bold text-zinc-50 mb-12 text-center">
          Add a new reservation
        </h2>
        <form className="flex flex-col gap-5 m-3 mt-0 forms-layout" onSubmit={handleSubmit(onSubmit)}>
          <div>
            <select
              id="citySelect"
              value={selectedCity}
              onChange={(e) => setSelectedCity(e.target.value)}
              placeholder="Select a City"
              className="forms-inputs bg-golden-yellow w-64 md:w-80"
              required
            >
              <option value="">Select a City</option>
              {cities.map((city) => (
                <option key={city.id} value={city.id}>
                  {city.name}
                </option>
              ))}
            </select>
          </div>
          <div>
            <select
              id="placeSelect"
              value={selectedPlace}
              onChange={(e) => setSelectedPlace(e.target.value)}
              placeholder="Select a Place"
              className="forms-inputs bg-golden-yellow w-64 md:w-80"
              required
            >
              <option value="">Select a Place</option>
              {places.map((place) => (
                <option key={place.id} value={place.id}>
                  {place.name}
                </option>
              ))}
            </select>
          </div>
          <div className="flex flex-col items-center justify-center w-full">
            <DatePicker
              id="datePicker"
              selected={selectedDate}
              onChange={(date) => setSelectedDate(date)}
              minDate={today}
              placeholderText="Select a Date"
              className="forms-inputs bg-golden-yellow w-64 md:w-80"
              autoComplete="off"
              required
            />
          </div>
          <button type="submit" className="forms-submit text-golden-yellow hover:bg-golden-yellow hover:text-white min-w-fit">
            Submit
          </button>
        </form>
      </div>
    </section>
  );
}

export default ReservationForm;
