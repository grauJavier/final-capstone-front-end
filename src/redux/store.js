import { configureStore, combineReducers } from '@reduxjs/toolkit';
import userReducer from './user/userSlice';
import { persistStore, persistReducer } from 'redux-persist';
import sessionStorage from 'redux-persist/lib/storage/session'; // defaults to localStorage for web
import { createTransform } from 'redux-persist';
import placesSlice from "./places/placesSlice";
import reservationReducer from './reservation_form/reservationSlice.js';
import myReservationsReducer from './my_reservations/myReservationsSlice.js';
import noticeAlertReducer from './notice_alert/noticeAlertSlice.js';

// Create a personalized transform.
const userTransform = createTransform(
  (inboundState) => {
    return {
      currentUser: inboundState.currentUser,
      token: inboundState.token
    };
  },

  (outboundState) => {
    return {
      ...outboundState
    };
  },
  // Apply this transform only to the user state.
  { whitelist: ['user'] }
);

const persistConfig = {
  key: 'root',
  storage: sessionStorage,
  whitelist: ['user'],
  transforms: [userTransform]
};

const reducer = combineReducers({
  noticeAlert: noticeAlertReducer,
  user: userReducer,
  placesSlice,
  reservation: reservationReducer,
  myReservations: myReservationsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false
    }),
});

const persistor = persistStore(store);

export { store, persistor };
