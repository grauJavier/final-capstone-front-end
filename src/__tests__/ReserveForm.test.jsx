import '@testing-library/jest-dom'; // For better DOM assertions
import { beforeEach, expect, describe, test } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import ReservationForm from '../components/reservation_form/ReservationForm';
import { store } from '../redux/store';

describe('reservation form', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route index path="/" element={<ReservationForm />} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  });

  test('should display the correct elements', async () => {
    const cityInput = screen.getByPlaceholderText('Select a City');
    const placeInput = screen.getByPlaceholderText('Select a Place');
    const dateInput = screen.getByPlaceholderText('Select a Date');

    const button = screen.getByRole('button', { name: 'Submit' });

    expect(cityInput).toBeInTheDocument();
    expect(placeInput).toBeInTheDocument();
    expect(dateInput).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });

  test('should render correctly', async () => {
    await waitFor(() => expect(ReservationForm).toMatchSnapshot());
  });
});
