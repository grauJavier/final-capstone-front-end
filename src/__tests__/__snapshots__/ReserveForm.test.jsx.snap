// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`reservation form > should render correctly 1`] = `[Function]`;
